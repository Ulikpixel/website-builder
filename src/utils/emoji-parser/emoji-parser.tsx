import React, { ReactNode } from 'react'
import Image from 'next/image'
import reactStringReplace from 'react-string-replace'

// Импорт SVG иконок из файлов
import HelloIcon from './icons/hello.svg'
import SmileIcon from './icons/smile.svg'

// Определение типа для карты эмодзи
type EmojiMap = { [key: string]: React.ReactNode }

export type InsertEmojisValue = string | ReactNode[]

// Статический объект с эмодзи в виде компонентов Image
const staticEmojiMap: EmojiMap = {
  hello: <Image src={HelloIcon} alt='Hello Icon' className='w-emoji h-emoji' />,
  smile: <Image src={SmileIcon} alt='Smile Icon' className='w-emoji h-emoji' />,
  // Добавьте другие эмодзи по необходимости
}

// Регулярное выражение для поиска смайликов вида !smile
const emojiCode = /(![a-zA-Z]+)/

/**
 * Функция заменяет ключевые слова в строке на соответствующие эмодзи из карты.
 * @param value - Входная строка с ключевыми словами, подлежащими замене на эмодзи.
 * @returns Строка с вставленными эмодзи.
 */
export function insertEmojis(value: InsertEmojisValue): React.ReactNode {
  return reactStringReplace(value, emojiCode, (match, i) => {
    // Получение ключевого слова из совпадения (например, !smile -> smile)
    const keyword = match.substring(1)

    // Возвращение React-элемента с смайликом и уникальным ключом
    return (
      <span key={`${i}-${Math.random()}`} className='inline-block relative top-px'>
        {staticEmojiMap[keyword]}
      </span>
    )
  })
}
