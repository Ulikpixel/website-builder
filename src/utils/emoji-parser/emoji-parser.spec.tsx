import React from 'react'
import { render } from '@testing-library/react'
import { insertEmojis, InsertEmojisValue } from './emoji-parser'

describe('insertEmojis', () => {
  it('should replace emojis in a string with corresponding React components', () => {
    const input: InsertEmojisValue = 'Hello !smile, how are you !hello?'
    const { container } = render(<>{insertEmojis(input)}</>)

    expect(container).toMatchSnapshot()
  })

  it('should handle an array of ReactNodes as input', () => {
    const input: InsertEmojisValue = ['Hello ', '!smile, how are you ', '!hello?']
    const { container } = render(<>{insertEmojis(input)}</>)

    expect(container).toMatchSnapshot()
  })

  it('should handle an empty string as input', () => {
    const input: InsertEmojisValue = ''
    const { container } = render(<>{insertEmojis(input)}</>)

    expect(container).toMatchSnapshot()
  })
})
