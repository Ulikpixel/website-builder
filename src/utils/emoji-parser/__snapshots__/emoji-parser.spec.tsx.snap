// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertEmojis should handle an array of ReactNodes as input 1`] = `
<div>
  Hello 
  <span
    class="inline-block relative top-px"
  >
    <img
      alt="Smile Icon"
      class="w-emoji h-emoji"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
  </span>
  , how are you 
  <span
    class="inline-block relative top-px"
  >
    <img
      alt="Hello Icon"
      class="w-emoji h-emoji"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
  </span>
  ?
</div>
`;

exports[`insertEmojis should handle an empty string as input 1`] = `<div />`;

exports[`insertEmojis should replace emojis in a string with corresponding React components 1`] = `
<div>
  Hello 
  <span
    class="inline-block relative top-px"
  >
    <img
      alt="Smile Icon"
      class="w-emoji h-emoji"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
  </span>
  , how are you 
  <span
    class="inline-block relative top-px"
  >
    <img
      alt="Hello Icon"
      class="w-emoji h-emoji"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
  </span>
  ?
</div>
`;
